app/controllers/api/v0/events_controller.rb

```diff
@@ -25,6 +27,16 @@ def convert_api_data_to_kafka_data(api_data:)
         data[:user_agent] = request.headers['User-Agent']
       end
 
+      # Split source_uri into parts, keeping most of them
```

Comment:
I think at some point Steve and I talked about generalizing this method and Steve made the good point to generalize after we have more than two kinds of data we're dealing with.  We're there now :-)

It is hard to test this code since it is in a private method.  You might try an inner class like so

```ruby
class Api::V0::EventsController < Api::V0::BaseController
  ...
  def create
    ...
    raw_kafka_data = KafkaData.new(api_data: event.data, controller: self)
    ...
  end

  class KafkaData
    def initialize(api_data:, controller:)
      @api_data = api_data.to_hash
      @controller = controller
      @data = {}

      translate_started_session
      translate_source_uri
      translate_uuids
    end
 
    def to_hash
      data
    end

    protected

    attr_reader :api_data, :controller, :data
  
    def translate_started_session
      return unless api_data[:type].include?('org.openstax.ec.started_session')
      data[:ip_address] = controller.request.remote_ip
      ...
    end

    def translate_source_uri
      # your URI(...) stuff here with any error checking
    end
  end
  ...
end
```

The advantage here is that you can hammer this inner class in specs, passing it a stubbed controller object that has only the little data needed here, e.g.

```ruby
require 'ostruct'
...
let(:remote_ip) { '192.168.1.1' }
let(:fake_controller) do
  OpenStruct.new(
    request: OpenStruct.new(
      remote_ip: remote_ip,
      ...
    )
  )
end

it 'translates the remote IP' do
  expect(described_class.new(api_data: {}, controller: fake_controller).to_h).to match(ip_address: remote_ip)
end
```

etc, etc

Reply:
could also do `KafkaData < Hash` which would eliminate need for `@data = {}`, `to_h`, etc

Reply:
Yeah, this is exactly where I expected some structural input. Will implement.

Reply:
Hmm, I appreciate the attempt to separate the input api_data from the output data, but I think Steve's modify-in-place strategy is probably the right one here: the bulk of fields will be passed on, with exceptions made clear in the translate_* methods. Otherwise, I'd need to add either a bulk copy of api_data => data, or knowledge of all the fields.